# Assignment 01. Exploratory data analysis with Pandas
## Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here")

In [17]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [18]:
data = pd.read_csv('../../data/titanic_train.csv',
                  index_col='PassengerId')

**First 5 rows**

In [19]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [20]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [21]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

**We can sort these people by Fare in descending order.**

In [22]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Let's create a new feature.**

In [83]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age > 60:
        return 3

In [84]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

**Another way is to do it with `apply`.**

In [25]:
data['Age_category'] = data['Age'].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men and 577 women VVV
- 479 men and 412 women
- 577 men and 314 women

In [33]:
data.groupby('Sex').count()

Survived  Pclass  Name  Age  SibSp  Parch  Ticket  Fare  Cabin  \
Sex                                                                      
female       314     314   314  261    314    314     314   314     97   
male         577     577   577  453    577    577     577   577    107   

        Embarked  Age_category  
Sex                             
female       312           261  
male         577           453

**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108 VVV
- 112
- 125

In [52]:
data.groupby(['Pclass', 'Sex']).count()

Survived  Name  Age  SibSp  Parch  Ticket  Fare  Cabin  \
Pclass Sex                                                              
1      female        94    94   85     94     94      94    94     81   
       male         122   122  101    122    122     122   122     95   
2      female        76    76   74     76     76      76    76     10   
       male         108   108   99    108    108     108   108      6   
3      female       144   144  102    144    144     144   144      6   
       male         347   347  253    347    347     347   347      6   

               Embarked  Age_category  
Pclass Sex                             
1      female        92            85  
       male         122           101  
2      female        76            74  
       male         108            99  
3      female       144           102  
       male         347           253

**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- median is  14.45, standard deviation is 49.69 VVV
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [60]:
(data['Fare'].median().round(2), data['Fare'].std().round(2))

(14.45, 49.69)

**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- No VVV


In [80]:
data.groupby('Survived').describe()

Pclass                                         Age         ...  Fare  \
          count  mean   std  min  25%  50%  75%  max  count   mean  ...   75%   
Survived                                                            ...         
0         549.0  2.53  0.74  1.0  2.0  3.0  3.0  3.0  424.0  30.63  ...  26.0   
1         342.0  1.95  0.86  1.0  1.0  2.0  3.0  3.0  290.0  28.34  ...  57.0   

                 Age_category                                       
             max        count  mean   std  min  25%  50%  75%  max  
Survived                                                            
0         263.00        424.0  1.53  0.62  1.0  1.0  1.0  2.0  3.0  
1         512.33        290.0  1.51  0.58  1.0  1.0  1.0  2.0  3.0  

[2 rows x 48 columns]

**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old VVV
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [142]:
data.groupby(['Age_category']).describe()

Survived                                      Pclass        ...  \
                count  mean   std  min  25%  50%  75%  max  count  mean  ...   
Age_category                                                             ...   
1.0             384.0  0.41  0.49  0.0  0.0  0.0  1.0  1.0  384.0  2.48  ...   
2.0             288.0  0.42  0.49  0.0  0.0  0.0  1.0  1.0  288.0  2.01  ...   
3.0              22.0  0.23  0.43  0.0  0.0  0.0  0.0  1.0   22.0  1.59  ...   

             Parch        Fare                                           \
               75%  max  count   mean    std   min    25%    50%    75%   
Age_category                                                              
1.0            1.0  4.0  384.0  29.21  43.54  0.00   7.92  13.00  29.12   
2.0            0.0  6.0  288.0  40.56  63.10  0.00   8.68  23.00  52.00   
3.0            0.0  4.0   22.0  41.37  54.78  6.24  10.50  26.55  45.79   

                      
                 max  
Age_category          
1.0           263.00  
2.0           512.33  
3.0           263.00  

[3 rows x 48 columns]

|**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women VVV

In [143]:
data.groupby(['Sex']).describe()

Survived                                      Pclass        ...   Fare  \
          count  mean   std  min  25%  50%  75%  max  count  mean  ...    75%   
Sex                                                                ...          
female    314.0  0.74  0.44  0.0  0.0  1.0  1.0  1.0  314.0  2.16  ...  55.00   
male      577.0  0.19  0.39  0.0  0.0  0.0  0.0  1.0  577.0  2.39  ...  26.55   

               Age_category                                       
           max        count  mean   std  min  25%  50%  75%  max  
Sex                                                               
female  512.33        254.0  1.43  0.52  1.0  1.0  1.0  2.0  3.0  
male    512.33        440.0  1.50  0.58  1.0  1.0  1.0  2.0  3.0  

[2 rows x 56 columns]

**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- William VVV
- John

In [210]:
def first_name(name):
    name = name[name.rindex(',')+2:]
    name = name[name.index('.')+2:]
    return name.split(' ')[0]

first_names = [first_name(name) for name in data.Name]

data['First_name'] = first_names

In [223]:
data[data['Sex'] == 'male'].groupby('First_name').count()

Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
First_name                                                                 
Abraham            1       1     1    1    1      1      1       1     1   
Achille            1       1     1    1    1      1      1       1     1   
Adola              1       1     1    1    1      1      1       1     1   
Adolf              1       1     1    1    1      1      1       1     1   
Adolphe            1       1     1    1    0      1      1       1     1   
...              ...     ...   ...  ...  ...    ...    ...     ...   ...   
William           35      35    35   35   31     35     35      35    35   
Wyckoff            1       1     1    1    1      1      1       1     1   
Yoto               1       1     1    1    1      1      1       1     1   
Youssef            1       1     1    1    0      1      1       1     1   
Yousseff           1       1     1    1    1      1      1       1     1   

            Cabin  Embarked  Age_category  
First_name                                 
Abraham         0         1             1  
Achille         0         1             1  
Adola           0         1             1  
Adolf           1         1             1  
Adolphe         1         1             0  
...           ...       ...           ...  
William         9        35            31  
Wyckoff         1         1             1  
Yoto            0         1             1  
Youssef         0         1             0  
Yousseff        0         1             1  

[287 rows x 12 columns]

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**
- On average, men of 1 class are older than 40 VVV
- On average, women of 1 class are older than 40
- Men of all classes are on average older than women of the same class VVV
- On average, passengers of the first class are older than those of the 2nd class who are older than passengers of the 3rd class VVV

In [226]:
data.groupby(['Sex', 'Pclass']).describe()

Survived                                         Age         \
                 count  mean   std  min  25%  50%  75%  max  count   mean   
Sex    Pclass                                                               
female 1          94.0  0.97  0.18  0.0  1.0  1.0  1.0  1.0   85.0  34.61   
       2          76.0  0.92  0.27  0.0  1.0  1.0  1.0  1.0   74.0  28.72   
       3         144.0  0.50  0.50  0.0  0.0  0.5  1.0  1.0  102.0  21.75   
male   1         122.0  0.37  0.48  0.0  0.0  0.0  1.0  1.0  101.0  41.28   
       2         108.0  0.16  0.37  0.0  0.0  0.0  0.0  1.0   99.0  30.74   
       3         347.0  0.14  0.34  0.0  0.0  0.0  0.0  1.0  253.0  26.51   

               ...    Fare         Age_category                             \
               ...     75%     max        count  mean   std  min  25%  50%   
Sex    Pclass  ...                                                           
female 1       ...  134.50  512.33         80.0  1.65  0.53  1.0  1.0  2.0   
       2       ...   26.06   65.00         72.0  1.43  0.50  1.0  1.0  1.0   
       3       ...   20.22   69.55        102.0  1.26  0.47  1.0  1.0  1.0   
male   1       ...   78.46  512.33         93.0  1.86  0.62  1.0  1.0  2.0   
       2       ...   26.00   73.50         96.0  1.54  0.56  1.0  1.0  2.0   
       3       ...   10.01   69.55        251.0  1.36  0.51  1.0  1.0  1.0   

                          
                75%  max  
Sex    Pclass             
female 1       2.00  3.0  
       2       2.00  2.0  
       3       1.75  3.0  
male   1       2.00  3.0  
       2       2.00  3.0  
       3       2.00  3.0  

[6 rows x 48 columns]

## Useful resources
* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-practice-analyzing-titanic-passengers) with a [solution](https://www.kaggle.com/kashnitsky/topic-1-practice-solution)
* Topic 1 "Exploratory Data Analysis with Pandas" as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)


Author: [Yury Kashnitsky](https://yorko.github.io). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.